### ISSUE TITLE
RFC: Sourcehut migration

### ISSUE THREAD

**Issue by Semisol** [tenacityteam/tenacity-legacy#155]
## This post is only advisory on the decision to migrate.

As discussed in #51, there has been thoughts about migrating to [Sourcehut](https://sr.ht/).
I would like to request opinion from the community.

**Please respond to this issue with the following comment format:**
```
I am in favor of option 1/2/3/4/5.

<insert reason>
```

Options for migration:
1. Full
We abandon GitHub and move to Sourcehut
2. Sourcehut primary, GitHub secondary
We will accept PRs and issues from GitHub, but mainly Sourcehut will be used. GH will be a mirror.
3. None
We will stay on GitHub and Sourcehut will not be used except for mailing lists and other misc stuff.
4. GitHub primary, Sourcehut secondary
We still continue to work as normal on GitHub, but Sourcehut can be contributed to.
Sourcehut will still be primary on mailing lists and other features GH doesn't have.
5. Something else
This option is for if you only want to move to another forge, partially or fully.

We recommend you read all the comments before responding.

**Comment by falkTX** [comment]
I am in favour of 2 (Sourcehut primary, GitHub secondary)
If that is not possible, 4 (GitHub primary, Sourcehut secondary)

While GitHub gets the most attention and users, it is not something I wish to stick around for the long-term.

We need to remember that those developers that are not okay with GitHub cannot come here and vote.
Having an alternative for them would be great.

**Comment by nbsp** [comment]
I am in favour of 2.

Speaking as a user, as I've left the team for a related reason.

As much as I'd like option 1, I understand it would be suicide for the project. It's just now gotten traction, deleting it would be the end.

Option 2 is the one I've been championing for since the start of the debacle. GitHub is a proprietary software forge, owned by a (subjectively) malicious company, and might be in violation of various viral licenses. Sound familiar?

The dislike toward sourcehut stems from 2 primary reasons:
1. You don't like Drew DeVault, for whatever reason, and do not wish to support him
2. The original "old" email-driven workflow is alien, and feels slower, and the UI is too simple

I'm not gonna argue with the former, but I can assure you that the UI being simple is a blessing; a recent [blog post](https://www.city17.xyz/declutter-social-coding/) I read goes into detail about that. It's also way more accessible. As I said, everyone online has email, not everyone has a GitHub account. Contributing on sourcehut does not require a sourcehut account.

Whatever the community decides on will be the right choice by definition; just my 2 cents.

**Comment by AnErrupTion** [comment]
I'm in favor of 3.

I would prefer option 4 but because I never used Sourcehut I would go with option 3, since GitHub has and will get you the most users.

**Comment by falkTX** [comment]
> I would prefer option 4 but because I never used Sourcehut I would go with option 3, since GitHub has and will get you the most users.

Why does it make a difference to you then? The idea with primary and secondary is that you would use one of them, not both.
So we can receive as many developers and contributions as possible.

**Comment by fossdd** [comment]
I am in favor of option 2 (alternativly 1, after that 4)

SourceHut is completly free and has a completly email-driven patch and discussion system like in many project like the linux kernel.

I would also like to see it completly on sourcehut that we don't have to look at open issues and PRs here on GitHub and can controll all withour email-client. But the most people also want to control their workflow over GitHub and we won't ignore these people.

I hope that many users come with and continue contributing. Hey: No account is needed, you can discuss and send patches in with your already (for GitHub required) E-Mail address. So there is no loss, just the people don't have to loss their fun contributing if there are 2 platforms or just one which is not so famous like GitHub is.

**Comment by fossdd** [comment]
> I would prefer option 4 but because I never used Sourcehut I would go with option 3, since GitHub has and will get you the most users.

Like I just said, you can do all with your email client and don't have to register or so. It's like the old-style linux workflow.

**Comment by nils-van-zuijlen** [comment]
I am in favor of option 2. (then 4)

As anyone with an email account can contribute to srht, and you have to have one to even use git, it seems to be a good solution.
Also, if anyone has issues with git-send-email, they will be able to submit PRs.

**Comment by Sciss** [comment]
Despite all valid criticism and problems of GH, I would always keep it as one platform/mirror of the project (thus __not option 1__). The next problem arises where you want users to interact, because you wouldn't want more than one ticket system, wiki, etc. Sourcehut doesn't even have a formal Wikipedia entry, to try to lift a project off the ground by moving to an obscure platform as first step, doesn't sound wise to me. And no, many people do not want to send and review git patches via e-mail.

**Comment by fossdd** [comment]
@Sciss Can you edit your post and add the `I am in favor of option ....`?

**Comment by Sciss** [comment]
Well, it is a request for comment, so there is my comment. Not sure this should be a vote by now.

**Comment by Be-ing** [comment]
Speaking from years of [experience](https://mixxx.org/news/2021-02-23-build-infrastructure-updates/) struggling with self hosted cross platform CI and setting up alternatives, I think that unless someone comes forward with a plan for cross platform CI besides GitHub Actions, this discussion is largely moot. Option 4 would be a reasonable compromise, but reviewers would need to diligently check CI results from GitHub Actions before merging on SourceHut.

**Comment by adrianheine** [comment]
I am in favor of option 2 (alternatively 5, 1)

5 being moving to another forge (probably gitea), possibly self-hosted.

**Comment by Semisol** [comment]
@adrianheine Added.

**Comment by JPcode05** [comment]
I am in favor of option 4. Most people use GitHub and discover the project via GitHub. Not everyone will play well with Sourcehut, especially git-send-email. **But my solution is to use _both_ at the same time.** If you prefer Sourcehut, use it. If you prefer GitHub, use that.

If another raid happens, in mailing lists, Sourcehut's servers may not be able to handle the load (since it's smaller than GitHub), and your Inbox will blow up just like Cookiengineer's did. Either your Email experience will be ruined because of the mailing list (and what if your mail server can't handle it?) or so many messages will be sent to spam that all of Sourcehut's future emails will be marked spam by your mail provider. Best example is Gmail, which most people use.

For general discussion (especially in forum help), people do not want to figure out how to sign up for Sourcehut to join a mailing list when all they need help with is how to use Tenacity, which in that case we will use Discourse. Developers, on the other hand, can sign up for either GitHub or Sourcehut to discuss development issues. Just be sure to get rid of the Discussion tab.

Edit: Just in case something wrong happens to GitHub, I do want this project to be on Sourcehut so that it's easier to abandon GitHub _when the time comes._

**Comment by Semisol** [comment]
I am in favor of 4 (then 2).

I currently do not want to confuse potential contributors on GH, but also provide people that do NOT want GH an alternative.

**Comment by JPcode05** [comment]
> I'm in favor of 3.
> 
> I would prefer option 4 but because I never used Sourcehut I would go with option 3, since GitHub has and will get you the most users.

I don't like 3. I prefer 4 over 3. (My vote comment is above.)

**Comment by fossdd** [comment]
Yes, understand @JPcode05 meaning. [Git is already federated & decentralized](https://drewdevault.com/2018/07/23/Git-is-already-distributed.html) and using pusing both to the two remotes would work. But we should have a main instance where we can people link to.

**Comment by Semisol** [comment]
> Yes, understand @JPcode05 meaning. [Git is already federated & decentralized](https://drewdevault.com/2018/07/23/Git-is-already-distributed.html) and using pusing both to the two remotes would work. But we should have a main instance where we can people link to.

That is what we decide as Primary will be.

**Comment by gregsskyles** [comment]
I'm in favour of 3.

I'd echo all of the reasons given above- it would be suicidal to migrate at this point as @SFR-git says, how to manage CI as @Be-ing says, and as several state, it would maximise the users, exposure and leverage what most people know. And as @Sciss suggests, I do not want to review git patches in email, and I'd worry about using something as obscure (to me) as SH is. I'm also not hearing what the advantage of SH is, other than its not M$ and the Linux kernel devs like it (are they going to contribute to this project?).

The email-driven issue management of the Audacity developer list was an ugly mess; GH issues are much superior. Maybe SH doesn't have this issue, but the email driven workflow comments have me worried.

All this being said, I could vote 2 or 4, if I knew anything about SH other than how to spell it and what the actual potential for interoperation is.

**Comment by Be-ing** [comment]
> maximise the users, exposure and leverage

FWIW, I think this is a bad reason to stick with GitHub. Cross platform CI is a compelling reason though.

**Comment by Sciss** [comment]
> I think this is a bad reason to stick with GitHub

Why is it a bad reason to want to lower the entry bar to new contributors and developers, some of which might not be as tech savvy as the core team? And exposure for a project that is not yet standing on its own feet, why is that bad? That doesn't have to mean we will forever stay on this platform. I think it doesn't preclude us from designing the process so that we are not locking our selves into this platform. At least GH has API to move things, e.g. migration to GitLab works pretty well.

**Comment by gregsskyles** [comment]
> > maximise the users, exposure and leverage
> 
> FWIW, I think this is a bad reason to stick with GitHub. Cross platform CI is a compelling reason though.

Yes, could be that's not what you want. Reading through issue #8 didn't really point one way or the other on this one.

**Comment by purplesyringa** [comment]
I am in favor of option 5 (Something else).

GitHub pros:
- Simple for beginners
- Almost everyone is familiar with it
- Lots of great built-in features such as milestones, projects, issues, PRs, CI, etc.

GitHub cons:
- Not FOSS
- Doesn't work well with LibreJS
- Too corporate / owned by Microsoft / whatever, if you see that as a problem

SourceHut pros:
- FOSS, works with LibreJS, etc.
- Great CI
- Simple, for some definition of simple

SourceHut cons:
- Not familiar to many people, requires yet another account, etc.
- Rather complicated for beginners: doesn't immediately show code tree, patches are mixed with issues, mailing lists, no labels/milestones/etc. for issues

Tenacity badly needs more contributors, so I think  switching to SourceHut at this point is not worth it; moreover, some of the problems of srht I listed apply to those who would like switch to a FOSS solution too. What about complete (or almost complete) synchronization of *three* platforms at once: GitHub, GitLab and SourceHut? GitHub would be useful for beginners (or as a landing page, because some people think that Git is synonymous to GitHub), SourceHut may be used by professionals and GitLab would be a great middle ground?

**Comment by nils-van-zuijlen** [comment]
> SourceHut cons:
> 
>    * Not familiar to many people, requires yet another account, etc.

Actually, as all of srht is email driven, contributors are not required to have an account to send patches (PR), todos (issues) nor an email to any list. That's a lower bar of entry than GitHub, GitLab or Gitea which all require you to have an account to do anything beside browsing.

**Comment by Semisol** [comment]
Three platforms feels like too big of a hassle.

**Comment by purplesyringa** [comment]
> Actually, as all of srht is email driven...

Which I believe is a con these days. GitHub, for instance:

- Differentiates issues from PRs, this is lost in email
- Supports formatting, which is not exactly easy to get in email
- Has great tools for code review, and all the comments are in one place, whereas in emails you have to read the whole thread
- Immediately shows the current state of the issue such as whether it is open or closed, issue labels, who is assigned, whether there are PRs that fix a bug if the issue is a bug report, etc. In emails, you have to grep the emails for 'state changed' or something.
- On GitHub and such stuff connected to any given repository lives at a well-known location: github.com/username/reponame, whereas your incoming mail box may be filled with spam, work mail, notifications and info about other repositories.
- The younger generation seldom uses email nowadays, which isn't a problem for us at the moment but may become one later on.

**Comment by falkTX** [comment]
> Tenacity badly needs more contributors

this is a little offtopic, but why do you think so?
if goal is to stay close to audacity, most of the work is going to be for initial setup, the rest is just trivial maintainer stuff.

PS: let's please not discuss git send-email vs github web approach cons vs pros here. this is a debate like many others (which build system is better, which language is better etc etc) that never end.
thanks.

**Comment by Semisol** [comment]
> > Tenacity badly needs more contributors
> 
> this is a little offtopic, but why do you think so?
> if goal is to stay close to audacity, most of the work is going to be for initial setup, the rest is just trivial maintainer stuff.
> 
> PS: let's please not discuss git send-email vs github web approach cons vs pros here. this is a debate like many others (which build system is better, which language is better etc etc) that never end.
> thanks.

they achieve the same task of submitting a contribution to be merged, so who cares too much?

**Comment by purplesyringa** [comment]
> this is a little offtopic, but why do you think so?

I thought this issue was for discussion about Sourcehut migration, uh? How's that not an argument against using Sourcehut?

**Comment by falkTX** [comment]
it is just going to be a heated debate that won't stop. people have different preferences and tastes.
You explained your reasons why you prefer a certain option. that is good for now.

**Comment by blackcrack** [comment]
> GitHub pros:
> 
>     * Simple for beginners
> 
>     * Almost everyone is familiar with it
> 
>     * Lots of great built-in features such as milestones, projects, issues, PRs, CI, etc.
> 
> 
> GitHub cons:
> 
>     * Not FOSS
> 
>     * Doesn't work well with LibreJS
> 
>     * Too corporate / owned by Microsoft / whatever, if you see that as a problem

so why not GitLab
https://about.gitlab.com/company/

not Microsoft, an Open Sourcefactory in general ,
and MS have already a OpensourcePortal.. GitHub,
i don't know if MS want buy Gitlab also.. ... so, well..

Gitlab have Open Source Partners :  
https://about.gitlab.com/partners/

there be a possible for find partners and many others...
to hold tenacity more healthy in the future

Option 5,

Primary Gitlab <-rsync -> Sourcehut 

this gives later more commercial possibility's at Gitlab
and therewith can the secured Sourcehut - sources be in able to make profit in any form from it.
therewith be with 2 places both tree's saved,

first it's possible for working with an free account on gitlab, later it's possible for let make more ,
if tenacity more supported from different parts..

Gitlab have Gui like Github and very easy for Beginners, Users and Extended Users and Programmers..
and i guess, also FOSS...

well, Gitlab is not only in the States..
https://about.gitlab.com/company/visiting/

check it out :  https://about.gitlab.com

so.. add a Option for Gitlab and Sourcehat

best

**Comment by purplesyringa** [comment]
> so why not GitLab

Which is exactly what I recommended, please read what others post, thanks.

**Comment by fitojb** [comment]
1 or 2, for me.

**Comment by blackcrack** [comment]
@imachug well, can i not give my Opinion ? be i here wrong ?

I am in favor of option 1 / 5
on Sourcehut do you not need a account, and you must not use an Gitlab account for working.. so what else ?
i have already account and it disturbs me not.. and right, Sourcehut be i am not familiar , 
this here is now where i have hear from..
Who was on Github can also work easy with Gitlab ..  and because this an rsync option make it possible for working in Sourcehut and supports with Gitlab the nonprogrammers .. the easy Users who want browse programs and want make a issue.. like in Github ... but in Gitlab

**Comment by Semisol** [comment]
> 1 or 2, for me.

Pick one and put the other(s) as "then x, y, z"

**Comment by Songtech-0912** [comment]
Current results (I'll keep tallying as the vote goes on):
- Option 1: 2.25
- Option 2: 6.5
- Option 3: 2
- Option 4: 4.5
- Option 5: 1.5
(Note - for second and third choices I added 0.5 votes and 0.25 votes respectively)

**Comment by fugidev** [comment]
I am in favor of option 4.

After reading all the arguments here as well as on matrix/libera, I personally think that for now, option 4 would be best fit. This way, new contributors are not scared off by the unfamiliar workflow on sr.ht, while those who want can simply submit patches via email.
I would possibly consider making sr.ht the primary forge in the future, but Github should be kept in some form because of 1. ready to use cross platform CI and 2. greater exposure to the public (which is important!).

**Comment by fossdd** [comment]
> > Actually, as all of srht is email driven...
> 
> Which I believe is a con these days.

I don't think so. Everyone has these days a email address (for example for GitHub,..). E-Mail is a federated protocol. Everyone can write messages to everyone else over E-Mail.

If someone doesn't want to write patches over emails, you can use github as secondary/primary.

> so why not GitLab

GitLab is not a good idea. It has A LOT of javascript which loads long. It requires Google Captcha, uses Cloudflare,... They aren't open source.

They are a for-profit-company. Of course it isn't directly a con, but you should think about it.

If you want to have a GitHub-like or GitLab-like platform I would suggest Codeberg. It is completly community-funded and uses the complete open-source Gitea. It has less JavaScript (does it have javascript?). And users already know it's design and workflow from Git(Hub/Lab).

> Primary Gitlab <-rsync -> Sourcehut

Using rsync is a bad idea. Git uses remotes. With remotes you can push it to multiple servers, and pull it, and have it integrated in your git workflow. rsync is too much and in my opinion not usefull.

**Comment by falkTX** [comment]
Let's please try to keep this on topic.
If you have a reason for why you prefer something in specific, write it on the post where you left your vote.
Thanks

**Comment by blackcrack** [comment]
@fossdd we have today Gightz Networks.. or they comes, Computers becomes ever more Power in the engines.. and the Partners of Gitlab be not MS supporter.. or ? and by the way Gitlab is more Prominent as .. ?Codeberg ? huhh again something new..
and as buy MS Github be a couples project goes to Gitlab.. how i become with.. the "Nonprofit" wellll : [Pricing](https://about.gitlab.com/pricing/)  
so less nonprofit is it also not..
Because rsync.. okey, i be no more on that stand of things, I must admit :) :grin: 

edit: @falkTX  Filipe, if this offtopic is fossdd posting also..

**Comment by Songtech-0912** [comment]
Current results (I'll keep tallying as the vote goes on, but this is the last comment I can do today, so it would be great if another contributor could step in):
- Option 1: 2.25
- Option 2: 6.5
- Option 3: 2
- Option 4: 5.5
- Option 5: 1.5
(Note - for second and third choices I added 0.5 votes and 0.25 votes respectively)

**Comment by unbeatable-101** [comment]
I am in favour of 5, if we are going to move the primary repo from github, I suggest [Codeberg](https://codeberg.org/). It will allow us to easily replicate most github functionality, so we could completely move away from it if that is what is wanted.

**Comment by fossdd** [comment]
## Latest results

- Option 1: 3.25
- Option 2: 8
- Option 3: 2
- Option 4: 6
- Option 5: 2.5

(Note - for second and third choices I added 0.5 votes and 0.25 votes respectively)

**Comment by ajayyy** [comment]
I am in favour of 4

**Comment by EchedelleLR** [comment]
I am in favour of option 2.

Option 1 if not.

**Comment by to7m** [comment]
I am in favour of option 5: Wait a while and do the vote again.

Sourcehut seems like a popular idea, but it's also in alpha. Maybe it would be best to delay any migration/mirroring effort, and make an agreement to hold the vote again once Sourcehut is more stable?

**Comment by Semisol** [comment]
> I am in favour of option 5: Wait a while and do the vote again.
> 
> Sourcehut seems like a popular idea, but it's also in alpha. Maybe it would be best to delay any migration/mirroring effort, and make an agreement to hold the vote again once Sourcehut is more stable?

Stability will take time and we are binding ourselves to GitHub more and more every day, which will make it harder to switch.

**Comment by fossdd** [comment]
> Sourcehut seems like a popular idea, but it's also in alpha. Maybe it would be best to delay any migration/mirroring effort, and make an agreement to hold the vote again once Sourcehut is more stable?

That could take a while. And maybe it's better to directly at the begin make a decision that the community can build on this platform.

**Comment by nils-van-zuijlen** [comment]
> > Sourcehut seems like a popular idea, but it's also in alpha. Maybe it would be best to delay any migration/mirroring effort, and make an agreement to hold the vote again once Sourcehut is more stable?
> 
> Stability will take time and we are binding ourselves to GitHub more and more every day, which will make it harder to switch.

Also, even if srht is in alpha, it's already the platform with the least downtime and the fastest pages. So it can be considered as a quite stable service. (APIs are changing and some parts need more polishing though)

**Comment by bl-ue** [comment]
If we want the most traction and activity, GitHub is definitely the way to go.  It's a very large community (in fact, about 50-100 user accounts are made on it _every minute_ as I recently manually determined).

A very many large projects use it too, and more are moving to it all the time.

Having a secondary system is definitely interesting, but it's perfectly possible for it to start getting complex, so let's keep that in mind.

I am in favor of **option 4**. (not changing my opt anymore 😆)

**Comment by nkeor** [comment]
I'm in favor of option 1.
If not, option 2.

**Comment by unbeatable-101** [comment]
> I am in favour of 5, if we are going to move the primary repo from github, I suggest [Codeberg](https://codeberg.org/). It will allow us to easily replicate most github functionality, so we could completely move away from it if that is what is wanted.

To expand on this, sourcehut isn’t as easy to learn how to use, so if we want to have more people contribute then it needs to be something similar to GitHub. As far as I know Codeberg fixes most of the concerns people have about GitHub. 

> It has less JavaScript (does it have javascript?).

It has javascript, but not much breaks if you disable it.

**Comment by caughtquick** [comment]
I'm in favor of option 2, though option 4 isn't bad.

**Comment by raavan13** [comment]
I would suggest Option 2

This will also be helpful for ppl who are not willing to leave GH and join SH for just this. I have accounts on both, so it doesn't affect me, but I would like his project to got to SH also.

**Comment by Seirdy** [comment]
Full disclosure: I am not an {Aud,Ten}acity contributor...yet. I've used it for a few years. Take this with a grain of salt. Despite that fact, I'm writing this partly because Tenacity could be an influential project, so other projects might follow its example.

Ranked choice (descending order): 2, 4, 5 (opt. 5 being a Gitea-based remote such as Codeberg or TildeGit). For the purposes of vote-tallying, consider this a vote for **option 2**.

If someone capable and trustworthy is able+willing to self-host, that's also an option.

Whatever option gets chosen, I'd like to suggest a [Hydra Hosting](https://seirdy.one/2020/11/18/git-workflow-1.html) approach. Hydra hosting (many remotes pushed/pulled to/from in parallel with a git alias) grants great hosting resilience with almost no additional effort, but it doesn't cover issue-tracking, support, or patch-submission; the choice of "primary" location should reflect these things.

As using GitHub comes with some ethical concerns for me and others in the Free Software community (e.g., much functionality requires enabling nonfree JS), and it requires creating an account (typical account-based service -> network effect -> vendor lock-in -> harder to leave if/when GH does something bad) to do anything beyond "read-only". It might be worth giving extra weight to ethical concerns like this because Tenacity was born out of very similar ethical concerns; the problems with Audacity apply to GitHub itself to a greater degree.

GitHub can be an option for existing users in the interests of making contribution more approachable, but I'd rather not make it the "default".

**Comment by mmmaisel** [comment]
I am in favor of option 3 or 4.

**Comment by Digit** [comment]
hi, audacity user for >17(?ish) years for all main stuff, will stick around (probably wont be much use though).  long live tenacity. 

I am in favor of option 1 or 5.

if not, prioritize away from gh.

5s: e.g. notabug.org, self hosted gitlab/gitea or whatever... idk, whatever is found best suitable for the project's ongoing freedom, after considering all the options.  something on federated githosting would seem ideal.  something activitypub backed?  idk, just an idea.  but if 1/sourcehut's good to go for tenacity, sourcehut.

**Comment by petersanchez** [comment]
Option 2 is the best choice here.

**Comment by bl-ue** [comment]
(To make this issue easier to read, I'm going to refrain from apply :+1: and :-1: on opinion comments -- and I sugggests others (e.g. JPcode05) to do the same -- it just makes the post a _lot_ harder to read with very little benefit)

**Comment by dhocker** [comment]
Here's a suggestion to ponder before you make a decision on what to do with the existing repo. 

The owner/leader of this project/repo should start a priority list of everything everyone wants to do. Order the priority list according to the value returned by an item (#1 yields the most ROI). Draw a cut-line at the point where items above the line are being worked on and items below the line are not being worked on (bandwidth limited). When a new item comes along, sort it into the list. Move the cut-line as necessary.

I don't have any idea what a priority list for this project might look like. But, I suspect that if you sort the item "move to a new source host" into such a priority list you might find that it is below the cut-line. I say that because the source host really doesn't return much value to the target user. It mostly returns value to the developers. To get such an item above the cut-line you would have to establish that there are other higher priority, more valuable items on the list that can't be accomplished without it.

If you make such a priority list you will likely find that there are many work items that have to be done just to move from the old Audacity repo to the new Tenacity repo. All of these items will meet what is known as the "must do" criteria, so they will be above the cut line (e.g. changing every occurrence of Audacity to Tenacity). Anything that does not meet the "must do" criteria will automatically get pushed down the list.

Just a thought...

Sorry...forgot to vote. This was really a recommendation for #3.

**Comment by Songtech-0912** [comment]
## Latest Results

Alright, I'm back to doing the tally:

- Option 1: 4.75
- Option 2: 11.5
- Option 3: 3
- Option 4: 7.5
- Option 5: 3.75

(Note - for second and third choices I added 0.5 votes and 0.25 votes respectively)

**Comment by Semisol** [comment]
@dhocker thanks for the idea, marked offtopic though. we will consider

**Comment by Flyingmana** [comment]
Voting for primary 4, secundary 3.

Reason: 
* Using Github is comfortable, keeping track of created Issues, done PRs, subscribe to single issues/changes to keep track of PR progress.
* Also good searchability and transparency.
* good CI integration
* providing many features related to delivery
* easier validating/Authentifikation of contributors by provided authentification with possible 2FA and easy validating of signed commits

Reason against sourcehut:
* using email to send patches is very uncomfortable and has lead to a few cases where I did need to resent patches, because of encoding issues
* I had bad experience, with submissions gotten lost and not receiving any feedback in the amount of other mails the maintainers had to handle.
* authorship of emails can easy be spoofed?
* it increases the hurdle for initial contributions high enough, that its likely to loose a lot of drive-through submissions. (for one of the projects I maintain, this drive through contributions make up more then 2/3 of all contributors https://cauldron.io/project/4125?from_date=2011-07-08&to_date=2021-07-08&tab=chaoss

**Comment by le-codechetif-blanc** [comment]
A self-hosted GitLab instance might also be a good idea- such as https://git.kiwifarms.net

**Comment by dhocker** [comment]
I vote for #3. See previous comment (that was removed as off-topic) about needs for a priority list to understand why I vote for #3.

**Comment by Semisol** [comment]
> I vote for #3. See previous comment (that was removed as off-topic) about needs for a priority list to understand why I vote for #3.

(not removed, but marked)

**Comment by bl-ue** [comment]
When we finally take all the votes and make a decision, we should probably take 1 and 2 together and 3 and 4 together as they're very similar.  1 and 2 prefer Sourcehut, 3 and 4 prefer GitHub.

**Comment by Songtech-0912** [comment]
Let's take in all the votes once we've gotten the votes of 40 contributors, and then I'll make another tally. I believe the votes of 40 contributors are close enough to represent the general wishes of the community

**Comment by Be-ing** [comment]
> Why is it a bad reason to want to lower the entry bar to new contributors and developers, some of which might not be as tech savvy as the core team?

I don't think trying to lower the barrier to entry to bad. I think it is important to pick a tool that is approachable for newcomers, so I think SourceHut is not a good choice. What I meant before is that I think making a decision to use a tool merely because everyone else does is not a good basis for that decision.

**Comment by Be-ing** [comment]
I think this voting is premature to make a final decision. The options haven't been discussed very thoroughly.

It seems few are really happy with GitHub, but we put up with it.

Considering nobody has presented any alternative proposals for cross platform CI, proposing to move exclusively to SourceHut is effectively proposing to drop macOS and Windows support, so I think that's out of the question.

As for GitLab, politically, I consider it only a little better than GitHub because it is owned by a private company heading for an IPO which has already [demonstrated a willingness to work with fascists](https://gitlab.com/gitlab-com/www-gitlab-com/-/merge_requests/30656) and has an [anti-community open core business model](https://media.libreplanet.org/u/libreplanet/m/why-i-forked-my-own-project-and-my-own-company-31c3/). Practically, [GitLab.com CI macOS Runners are in beta](https://gitlab.com/gitlab-org/gitlab/-/issues/269756) and so are [Windows runners](https://docs.gitlab.com/ee/ci/runners/build_cloud/windows_build_cloud.html).

Gitea I think deserves more discussion. Its design is very similar to GitHub but it is a community project, not some corporate open core thing like GitLab. AFAIK the only cross platform CI that works with Gitea is AppVeyor, which only provides one job at a time for free. Speaking from experience, that's not adequate for an active project. GitHub Actions is much more generous with computing resources, allowing up to 20 concurrent jobs (or max 5 concurrent macOS jobs). Some [compatibility with GitHub Actions](https://github.com/go-gitea/gitea/issues/13539) is on the Gitea roadmap, so I think it could be feasible in the future to move from GitHub fully to Gitea.

**Comment by SethFalco** [comment]
I favor option 5, then 4.

I think [GitLab](https://about.gitlab.com/) would be nice to use:
* Stable
* Continue Integration
* Self-Hostable
* Open-Source (Unfortunately, it's open-core though, similarly to Bitwarden. So, not _100%_ open-source like other solutions.)

Overall, I'm not too fussed where it is so long as people can contribute easily, though.

I do think it'd be nice to keep the project on GitHub one way or another (primarily, just for PRs, or even as a read-only mirror) just so it's easy to find for people that look for projects here. (ie via [topics](https://github.com/topics))

**Comment by Be-ing** [comment]
Provided that the beta macOS and Windows shared runners work well, I think GitLab.com would be the best option at this point. That leaves open the possibility of moving to self-hosted GitLab or Gitea (self-hosted, Codeberg, or some other host) in the future.

**Comment by Songtech-0912** [comment]
> I think this voting is premature to make a final decision. The options haven't been discussed very thoroughly.

I think most contributors agree with this sentiment. However, in the absence of a true governance structure, voting is a good way to find the general positions of the community - after the vote is over (at 40 contributors) I can write a full breakdown and summary of the vote results, including ideas placed forward by the community. Overall, I believe we should take a max-benefit least-risk approach to this, basically we compare each of the expressed options and then find the one which gives us the maximum net benefit without exceeding an agreed-upon risk/issue threshold.

**Comment by EchedelleLR** [comment]
I ask here.

Is not an hybrid solution possible?

In the beginning, I voted option 2 thinking on this.

Part of the CI at SourceHut and part at GitHub and allowed contributions in both sides.

You can maintain contributions here and allow the easy migration to a fully Free Software solution.

**Comment by Be-ing** [comment]
> Part of the CI at SourceHut and part at GitHub and allowed contributions in both sides.

Not really. CI should tell you if a merge/pull request would break builds or fail tests *before* it is merged. SourceHut cannot tell you if code changes would break Windows or macOS builds.

**Comment by Be-ing** [comment]
> voting is a good way to find the general positions of the community

I agree that the voting has been helpful for getting a general idea of where the community is, but I don't think the voting should be taking as definitive at this point. I get the impression many who voted for SourceHut were more voting for "not GitHub" than really voting *for* SourceHut.

**Comment by EchedelleLR** [comment]
> Not really. CI should tell you if a merge/pull request would break builds or fail tests before it is merged. SourceHut cannot tell you if code changes would break Windows or macOS builds.

Well, this is not what I am used to do, given some projects in which I saw direct commits and the builds runs after it. If this is a behaviour that can be used I don't see any problem.

**Comment by Be-ing** [comment]
>  direct commits and the builds runs after it

All code should be reviewed *before* it is merged to the upstream branch unless you're working alone.

**Comment by EchedelleLR** [comment]
What would be the issue working with a development branch for it?

Then, periodical merges to the main branch. Some projects (I don't have anything in my mind right now but I remember to see it) work like that.

**Comment by Be-ing** [comment]
This is getting off topic. The more important point is that SourceHut's CI does not support Windows or macOS.

**Comment by EchedelleLR** [comment]
> This is getting off topic. The more important point is that SourceHut's CI does not support Windows or macOS.

I am trying to solve the background issues after it supporting the idea of the hybrid environment.

It is not offtopic if I am just trying to solve any issue you submit to the idea.

**Comment by emabrey** [comment]
I am in favor of option 4.

**Comment by unbeatable-101** [comment]
> > voting is a good way to find the general positions of the community
> 
> I agree that the voting has been helpful for getting a general idea of where the community is, but I don't think the voting should be taking as definitive at this point. I get the impression many who voted for SourceHut were more voting for "not GitHub" than really voting _for_ SourceHut.

That is what I got from this too, an actual vote should give more options other than GitHub and SourceHut. Also no matter what we do I think we should leave github up as a mirror, even if we don’t accept contributions to it (Like https://github.com/apple/darwin-xnu/).

**Comment by davidak** [comment]
I am in favor of option 5.

I think Codeberg would be the best option as it has similar workflow to Github that most developers are used to without the bloated nature of GitLab.

- They don't offer CI yet, but are working on it. A self-hosted CI can be used already (I tested that successfully).
- They are testing GitHub login right now, so every GitHub user can easily contribute. GitLab also considered.
- It is a community maintained non-profit based in Germany (strong privacy laws).

I'm against Sourcehut because i think the e-mail workflow is not user friendly (i tried once and it was hard to setup). It feels very archaic to me and is comparable to IRC. I think people that still prefer such technology from the 90s are also using window managers and CLI tools. Nothing against them, but i think it's not for the majority of people and projects should be accessible to everyone.

**Comment by falkTX** [comment]
If we use mingw, sourcehut can compile windows binaries just fine.
There are some interesting projects out there to run macOS under qemu/kvm under docker, worth checking at some point.

We have to be aware that the tool picked needs to be good for new comers yes, but it should cater to the wishes/needs of a core team too because those are the ones that are going to maintain the thing..

**Comment by fossdd** [comment]
> If we use mingw, sourcehut can compile windows binaries just fine.

Isn't that for have a GNU environment on Windows? And not for Windows under Linux. Maybe building for Windows work with Wine

**Comment by falkTX** [comment]
It can do both. cross-compile from linux and give unix-like environment on windows.
The final binaries work on stock windows, I built audacity like that before.

**Comment by Be-ing** [comment]
If official builds rely on cross compiling, I think it is unlikely that there will be macOS or Windows contributors, which makes it unlikely that platform-specific bugs would get fixed.

**Comment by falkTX** [comment]
Not my call to make, but on Windows 10 and 11 running Linux CLI stuff is pretty simple now.
And mingw has native windows tools too, no need for VMs and cross-compilation.

**Comment by Be-ing** [comment]
If we switch to cross compiling for official builds, someone needs to figure out how to maintainably make the same build system usable on macOS and Windows and thoroughly document how to builds on macOS and Windows. Otherwise I have little confidence that OS-specific issues will get fixed.

**Comment by Flyingmana** [comment]
Lets try to move the discussion about Build environments out here for better visibility.
I created one for Windows here: https://github.com/tenacityteam/tenacity/discussions/183
I think OSX should get a separate topic, even as some approaches may overlap

**Comment by Be-ing** [comment]
Then let's put aside this discussion until we decide on that because it affects what options are available for this discussion.

**Comment by nova-nowiz** [comment]
I am in favour of 5.

My preferred platform would be Codeberg (thanks @davidak for your comment). It is based on gitea and is hosted in the EU.
The UI is super clear like github so it would be very easy to switch from github to codeberg.

I also think the switch can be done later. This repository and organization are still very young, so I think this issue is not a priority right now. Furthermore, it will increase friction for contributors and require a lot of work in terms of setup (finding solutions for the CI is the more pressing one).

My guess is that the smartest thing to do in this case is to make the switch in steps.
First, have github as the primary platform but setup and improve the alternative platform as time passes.
Then at some point when all roadblocks and friction points have been cleared and the alternative platform is now as or more functional/usable than github, make a public announcement stating that the former secondary platform is now the primary one.

This hybrid solution would provide time to setup the alternative platform and also testing of the alternative platform. When it is made stable and usable as the main one, switch.
The switch would in this case only be political or to gather the community in one place as the two platforms would be in fact functional.

**Comment by purplesyringa** [comment]
> finding solutions for the CI is the more pressing one

I believe repository hosting and CI can be decided on separately. If sourcehut, for instance, only supported Linux CI, it wouldn't be very difficult to write a script to run Windows and macOS CI on GitHub or GitLab or something else and then show the logs on sourcehut (I've done something similar before)

**Comment by Be-ing** [comment]
> If we switch to cross compiling for official builds, someone needs to figure out how to maintainably make the same build system usable on macOS and Windows and thoroughly document how to builds on macOS and Windows. Otherwise I have little confidence that OS-specific issues will get fixed.

If we can get this working so we only need Linux CI runners, my preferences would be Gitea, then GitLab, then SourceHut.

**Comment by Be-ing** [comment]
@falkTX if you can open a PR to get cross compiling working on a GitHub Actions Linux runner, I think that would be a great proof of concept to move this discussion forward.

**Comment by falkTX** [comment]
At some point will do. Still need to read stuff about github actions, I never used them before (only modified someone else's configs)

**Comment by NotAProton** [comment]
I am in favor of 4 then 2
Github requires an account and accepting of their terms & privacy policy.

**Comment by RoaddogLabs** [comment]
How does changing the source control benefit the project?  If the goal of the project is to get as many contributors and expose the project to the largest possible audience then right here is the obvious choice.  So far I've not seen one compelling technical reason to have two methods let alone leave here altogether.  The reality is that for the fork to succeed it needs to be here in a manner where it isn't treated as second to another solution.

It's far from a done deal this fork will come close to the penetration of Audacity.  Or even ship a point release.  Too soon to tell.  For that to happen there needs to be consistency in all things related to the project.  Part of that consistency is retaining a presence here.  It has nothing to do with who owns the platform, whether it's open source or any other altruistic leanings.  It has to do with traction and visibility.  Getting the best audio editing program possible should be the goal.  The tools and workflow to make that happen are secondary in this case.  There is more downside to moving than there is to staying.

**Comment by caughtquick** [comment]
> How does changing the source control benefit the project?  If the goal of the project is to get as many contributors and expose the project to the largest possible audience then right here is the obvious choice.  So far I've not seen one compelling technical reason to have two methods let alone leave here altogether.  The reality is that for the fork to succeed it needs to be here in a manner where it isn't treated as second to another solution.
> 
> 
> 
> It's far from a done deal this fork will come close to the penetration of Audacity.  Or even ship a point release.  Too soon to tell.  For that to happen there needs to be consistency in all things related to the project.  Part of that consistency is retaining a presence here.  It has nothing to do with who owns the platform, whether it's open source or any other altruistic leanings.  It has to do with traction and visibility.  Getting the best audio editing program possible should be the goal.  The tools and workflow to make that happen are secondary in this case.  There is more downside to moving than there is to staying.

That's my opinion too, I think we should stay on GitHub until this project is known for more than the GitHub repository.

**Comment by Be-ing** [comment]
> The reality is that for the fork to succeed it needs to be here in a manner where it isn't treated as second to another solution.

This is an obviously false assertion. GitHub does not have a monopoly on software development.

**Comment by 12people** [comment]
I am in favor of option 5. Specifically, moving to Gitlab and mirroring code only to Github.

As some have mentioned above, Sourcehut is only in alpha state, which makes it a risky choice for a larger long-term project. More importantly, it's not particularly user friendly and could potentially discourage people from contributing.

We can learn from [GNOME](https://securityonline.info/gnome-move-to-gitlab/)'s, [Freedesktop.org](https://securityonline.info/following-gnome-freedesktop-org-migrates-to-gitlab/)'s and [KDE's](https://www.world-today-news.com/code-hosting-kde-migrates-to-gitlab/) migrations here. I don't know about the others, but GNOME migrated because their previous stack was not very contributor friendly and was also getting harder to maintain. Sure enough, once they migrated, they saw a rise in contributions — showing that the user-friendliness of a platform really does matter.

Anyway, if Sourcehut is chosen, I'd be in favor of having a window (3 months?) where Github would still be primary and Sourcehut secondary. That way, developers could test out Sourcehut and see if it really meets their needs.

**Comment by blackcrack** [comment]
please Vote and not only Reason 
>
>
>I am in favor of option 1/2/3/4/5.
> \<insert reason>

and any can have his reason, but it must justified
if like, thump's up or not it's no matter and can f... the same imho, so please, give a clean VOTE 
at me is  5th or 2 and read the ..beep.. Topic on top if you are write something ... 
Gitlab and Sourcehut at me, but this have i already say's more top why..
in short: Independent of MS (Github belongs MS) and the States ...

(that comes out when you hide something)
( i want not directly goes offtopic, with a new post, but @Be-ing , hey DJ, Gitlab supports the most systems, include WinNT and Sourcehat too so far i am know.. so, imho must first goes independents and support Linuxbin's then thinking about WinNT and Mac is a Linuxsystem with own restricted driver in short, so.. what about "cross compiling" and if need it's able to use external compiling engins DJ.. i know you have wrote Mixxx ... )

**Comment by Be-ing** [comment]
It is too early to vote on this. We have work to do exploring technical options, namely cross compiling.

**Comment by Songtech-0912** [comment]
@Be-ing As stated previously, this is an advisory that will be taken into consideration for the final vote. Anyways, 2 more contributors before the final vote is taken and I do the tally! I am voting for **option 4**.

**Comment by fossdd** [comment]
> It is too early to vote on this. We have work to do exploring technical options, namely cross compiling.

But If we wait longer, the migration would getting harder and harder.

**Comment by bl-ue** [comment]
Also, we don't have enough normal users and potential developers in the voting.  Most of the people voting for 1/2/5 have moved away from GitHub, and lots of them say so on their GH profiles.  😕

**Comment by ghost** [comment]
If the opinon of users were requested too, I want to express mine. If not, you can dismiss my vote.

I'm in favour of (3), remaining on Github. Reasons are as follows:

- Sourcehut is a paid-for platform (tho currently it's free in its alpha IIUC). There's an exception for those who use it only to contribute but that might change and/or be confusing to newcomers.
- Sourcehut is run mostly by a single person. If a non-mainstream code forge is to be used, I feel like it's more appropriate to opt for a platform that is collectively community owned.

As for (5), I think there's not much that GH offers that GL doesn't. But IMHO postponing a move until after the fork is in some OS repos, a new website is made, and some traction is gained is wiser. As @Be-ing says, GH shouldn't have a monopoly on hosting FOSS, but I don't think there's harm in taking advantage of the existing contribution routes that the wider community is used to while the fork establishes itself.

Anyways, my 2c. Thank you so much for your perseverance on maintaining this fork and doing the right thing, despite a bunch of sad stuff we had to put up with.

**Comment by kubo6472** [comment]
I would do it in [this way](https://github.com/cloudrac3r/bibliogram#bibliogram-has-moved-to-sourcehut), if debating about whether or not move to sourcehut. Keep in mind, that you can host sourcehut software yourself. However, for performance and ease-of-use reasons, I much prefer [gogs](https://gogs.io) as a self-hosted product.

**Comment by Be-ing** [comment]
> But If we wait longer, the migration would getting harder and harder.

I agree that we shouldn't wait for some arbitrary time to switch. I think we should switch when we have a solution figured out for macOS and Windows builds.

**Comment by kubo6472** [comment]
> I agree that we shouldn't wait for some arbitrary time to switch. I think we should switch when we have a solution figured out for macOS and Windows builds.

I think you can build at least Windows binaries in [docker containers](https://hub.docker.com/_/microsoft-windows-servercore).

**Comment by Be-ing** [comment]
That does not help:
> Windows requires the host OS version to match the container OS version.

**Comment by fossdd** [comment]
> I would do it in [this way](https://github.com/cloudrac3r/bibliogram#bibliogram-has-moved-to-sourcehut), if debating about whether or not move to sourcehut.

That would option 1.

> Keep in mind, that you can host sourcehut software yourself.

Of course, but there is not a big reason. SourceHut doesn't require a account, all contributors can contribute with only having a email address. It is maybe a bit much for "just" a audacity fork, sr.ht already has it's servers.

> However, for performance and ease-of-use reasons, I much prefer [gogs](https://gogs.io) as a self-hosted product.

Gogs (and also the fork, gitea) require a account for contributors if they want to submit a patch.

> If the opinon of users were requested too, I want to express mine. If not, you can dismiss my vote.

The team would, i guess, allow everyone to vote. Users, Developers, and so on, everybody is welcome as it's a free and open source community-driven project.

**Comment by Semisol** [comment]
I have started tallying. The issue will be locked to collaborators until the results are out.

**Comment by Semisol** [comment]
Please note that only clear (`I am in favor of x (optional: then y, z)`) choices will be included in the tally.

**Comment by Semisol** [comment]
Using [instant runoff voting](https://en.wikipedia.org/wiki/Alternative_vote) (like previously done in the name private vote), the winner is:
## Sourcehut primary, Github secondary.

<details>
<summary>Ballot data</summary>
BLT file
<pre>
5 1
1 2 4 - 0
1 2 - - 0
1 3 - - 0
1 2 1 4 0
1 2 4 - 0
1 2 5 1 0
1 4 - - 0
1 4 2 - 0
1 3 - - 0
1 5 - - 0
1 1=2 - - 0
1 1=5 - - 0
1 4 - - 0
1 5 - - 0
1 4 - - 0
1 2 1 - 0
1 5 - - 0
1 4 - - 0
1 1 2 - 0
1 2 - - 0
1 2 - - 0
1 2 4 5 0
1 3=4 - - 0
1 1=5 - - 0
1 2 - - 0
1 4 3 - 0
1 3 - - 0
1 5 4 - 0
1 4 - - 0
1 5 - - 0
1 5 - - 0
1 4 2 - 0
1 5 - - 0
1 4 - - 0
1 3 - - 0
1 2 - - 0
0
"Sourcehut"
"Sourcehut primary Github secondary"
"Github"
"Github primary Sourcehut secondary"
"Other"
"Vote for forge selection"
</pre>

The raw ballots:
<pre>
2 4 -
2 - -
3 - -
2 1 4
2 4 -
2 5 1
4 - -
4 2 -
3 - -
5 - -
1=2 - -
1=5 - -
4 - -
5 - -
4 - -
2 1 -
5 - -
4 - -
1 2 -
2 - -
2 - -
2 4 5
3=4 - -
1=5 - -
2 - -
4 3 -
3 - -
5 4 -
4 - -
5 - -
5 - -
4 2 -
5 - -
4 - -
3 - -
2 - -
</pre>
(- = unspecified, `a=b` means both were selected)


</details>

OpenSTV logs:
```
OpenSTV version 1.7 (http://www.OpenSTV.org/)

Suggested donation for using OpenSTV for an election is $50.  Please go to
http://www.OpenSTV.org/donate to donate via PayPal, Google Checkout, or
Amazon Payments.

Certified election reports are also available.  Please go to
http://www.openstv.org/certified-reports for more information.

Loading ballots from file test.blt.
Ballot file contains 5 candidates and 36 ballots.
No candidates have withdrawn.
Ballot file contains 32 non-empty ballots.

Counting votes for Vote for forge selection using Instant Runoff Voting.
5 candidates running for 1 seat.

 R|Sourcehut  |Sourcehut p|Github     |Github prim|Other      |Exhausted
  |           |rimary Gith|           |ary Sourceh|           |
  |           |ub secondar|           |ut secondar|           |
  |           |y          |           |y          |           |
==========================================================================
 1|          1|         11|          4|          9|          7|          0
  |-----------------------------------------------------------------------
  | Count of first choices.
==========================================================================
 2|           |         12|          4|          9|          7|          0
  |-----------------------------------------------------------------------
  | Count after eliminating Sourcehut and transferring votes.
==========================================================================
 3|           |         12|           |          9|          7|          4
  |-----------------------------------------------------------------------
  | Count after eliminating Github and transferring votes.
==========================================================================
 4|           |         12|           |         10|           |         10
  |-----------------------------------------------------------------------
  | Count after eliminating Other and transferring votes. Candidate
  | Sourcehut primary Github secondary is elected.

Winner is Sourcehut primary Github secondary.
```

**Comment by EchedelleLR** [comment]
@Semisol why did you use propietary software for this?

For example, I suggested https://poll.disroot.org specifically to avoid this...

**Comment by EchedelleLR** [comment]
Ah okay, that was the last FLOSS version. I read.

Sorry.

**Comment by Semisol** [comment]
> Ah okay, that was the last FLOSS version. I read.
> 
> Sorry.

I don't want to use the web OpaVote again after trying it (no source code annoys me too).

**Comment by Semisol** [comment]
I have locked the conversation to freeze the votes.
Thanks for participating, if there are any concerns 
send an e-mail to the `~tenacity/tenacity-discuss@lists.sr.ht` mailing
list or open a discussion.
